// aicp-tracker/src/log-parser.js
'use strict';

const fs    = require('fs');
const state = require('./state');

const WRITE_TOOLS = new Set(['Edit', 'Write', 'NotebookEdit', 'MultiEdit']);

function extractTasksFromBranch(branch) {
  if (!branch || typeof branch !== 'string') return [];
  const matches = branch.match(/[A-Za-z][A-Za-z0-9]*-\d+/g) || [];
  return [...new Set(matches.map(m => m.toUpperCase()))];
}

// Sum usage fields from an array of assistant entries
function sumUsage(entries) {
  let input = 0, cc = 0, cr = 0, out = 0;
  let ephUser = 0, ephAsst = 0, ephTool = 0;
  let webSearch = 0, webFetch = 0;
  let serviceTier = null, speed = null;
  const models = new Set();
  const uuids = [];

  for (const e of entries) {
    const u = e.message?.usage || {};
    input    += u.input_tokens                     || 0;
    cc       += u.cache_creation_input_tokens      || 0;
    cr       += u.cache_read_input_tokens          || 0;
    out      += u.output_tokens                    || 0;
    ephUser  += u.ephemeral_user_input_tokens      || 0;
    ephAsst  += u.ephemeral_assistant_input_tokens || 0;
    ephTool  += u.ephemeral_tool_input_tokens      || 0;
    if (!serviceTier && u.service_tier) serviceTier = u.service_tier;
    if (!speed && u.speed) speed = u.speed;

    const content = e.message?.content;
    if (Array.isArray(content)) {
      for (const block of content) {
        if (block.type !== 'tool_use') continue;
        if (block.name === 'web_search' || block.name === 'WebSearch') webSearch++;
        if (block.name === 'web_fetch'  || block.name === 'WebFetch')  webFetch++;
      }
    }
    if (e.message?.model) models.add(e.message.model);
    if (e.uuid) uuids.push(e.uuid);
  }
  return { input, cc, cr, out, ephUser, ephAsst, ephTool, webSearch, webFetch, serviceTier, speed, models, uuids };
}

function makeRecord(first, rootUuid, usage, filePaths) {
  const { input, cc, cr, out, ephUser, ephAsst, ephTool, webSearch, webFetch, serviceTier, speed, models, uuids } = usage;
  return {
    sessionId:                        first.sessionId  || null,
    uuid:                             rootUuid,
    parentUuid:                       first.parentUuid || null,
    timestamp:                        first.timestamp  || new Date().toISOString(),
    gitBranch:                        first.gitBranch  || null,
    task_from_git_branch:             extractTasksFromBranch(first.gitBranch),
    model:                            [...models].join(',') || null,
    input_tokens:                     input,
    cache_creation_input_tokens:      cc,
    cache_read_input_tokens:          cr,
    output_tokens:                    out,
    ephemeral_user_input_tokens:      ephUser,
    ephemeral_assistant_input_tokens: ephAsst,
    ephemeral_tool_input_tokens:      ephTool,
    web_search_requests:              webSearch,
    web_fetch_requests:               webFetch,
    file_paths:                       filePaths,
    enterprise_usd_per_token:         null,
    service_tier:                     serviceTier,
    speed:                            speed || (cc > 0 ? 'fast' : 'normal'),
    trace_uuids:                      uuids.join(','),
  };
}

function parseNewLines(filePath) {
  const carryOverPromptId = state.getLastPromptId(filePath);

  const offset = state.getOffset(filePath);
  let stat;
  try { stat = fs.statSync(filePath); } catch { return []; }
  if (stat.size <= offset) return [];

  const fd  = fs.openSync(filePath, 'r');
  const buf = Buffer.alloc(stat.size - offset);
  const read = fs.readSync(fd, buf, 0, buf.length, offset);
  fs.closeSync(fd);
  state.setOffset(filePath, offset + read);

  // ── Parse all entries ─────────────────────────────────────────────────────
  // message.id deduplication: multi-block API responses (thinking + text + tool_use)
  // share the same message.id but have different UUIDs. Only the first occurrence
  // carries the usage fields that should be counted; subsequent lines are duplicates.
  const byUuid     = new Map();
  const seenMsgIds = new Set();

  for (const raw of buf.slice(0, read).toString('utf8').split('\n')) {
    const line = raw.trim();
    if (!line) continue;
    let entry;
    try { entry = JSON.parse(line); } catch { continue; }
    if (!entry.uuid) continue;

    const msgId = entry.message?.id;
    if (msgId) {
      if (seenMsgIds.has(msgId)) continue;
      seenMsgIds.add(msgId);
    }
    byUuid.set(entry.uuid, entry);
  }

  // ── Build promptId index for all user entries in this batch ───────────────
  // Every user entry (real prompt or tool-result) with a `promptId` field gets
  // registered. The promptId is a turn-level ID shared by the initiating user
  // message and all tool-result user entries that follow in the same turn.
  // For old entries without the field, real prompts (text content) use their own uuid.
  //
  // promptIdByUuid maps: entry.uuid → promptId (turn root identifier)
  const promptIdByUuid = new Map();
  let latestPromptId   = carryOverPromptId;

  for (const entry of byUuid.values()) {
    const isUser = entry.type === 'user' || entry.message?.role === 'user';
    if (!isUser) continue;

    if (entry.promptId) {
      promptIdByUuid.set(entry.uuid, entry.promptId);
      latestPromptId = entry.promptId;
    } else {
      // No promptId field — check if it's a real user text message (not tool-result only)
      const content = entry.message?.content ?? entry.content;
      const hasText = Array.isArray(content)
        ? content.some(c => c.type === 'text')
        : (typeof content === 'string' && content.length > 0);
      if (hasText) {
        promptIdByUuid.set(entry.uuid, entry.uuid);
        latestPromptId = entry.uuid;
      }
    }
  }

  // ── Resolve promptId for an assistant entry ───────────────────────────────
  // Walk the parentUuid chain upward. The first ancestor found in promptIdByUuid
  // determines which turn this assistant entry belongs to.
  // If the chain exits the batch (parent not in byUuid), fall back to the
  // carry-over from the previous batch.
  function resolvePromptId(entry) {
    const visited = new Set();
    let cur = entry;
    while (cur && !visited.has(cur.uuid)) {
      visited.add(cur.uuid);
      const parentId = cur.parentUuid;
      if (!parentId) break;
      if (promptIdByUuid.has(parentId)) return promptIdByUuid.get(parentId);
      const parent = byUuid.get(parentId);
      if (!parent) return carryOverPromptId || entry.uuid; // batch boundary
      cur = parent;
    }
    return carryOverPromptId || entry.uuid;
  }

  // If carryOverPromptId was re-introduced by a user entry in this batch (e.g. the
  // real user message itself landed here), treat it as a normal in-batch group so
  // the uuid stays as-is.  Only suffix when the carry-over was the ONLY way this
  // promptId was seen (i.e. no user entry in this batch declares that promptId).
  const carryOverSeenInBatch = carryOverPromptId &&
    [...promptIdByUuid.values()].includes(carryOverPromptId);

  // ── Group assistant turns by promptId ─────────────────────────────────────
  const promptGroups = new Map(); // promptId → assistant entries[]

  for (const entry of byUuid.values()) {
    const isAssistant = entry.type === 'assistant' || entry.message?.role === 'assistant';
    if (!isAssistant || !entry.message?.usage) continue;
    const pid = resolvePromptId(entry);
    if (!promptGroups.has(pid)) promptGroups.set(pid, []);
    promptGroups.get(pid).push(entry);
  }

  // Persist the latest promptId so the next batch can attribute orphaned entries
  if (latestPromptId) state.setLastPromptId(filePath, latestPromptId);

  // ── Per prompt: bucket turns by file anchor, emit one record per file ─────
  const records = [];

  for (const [promptId, turns] of promptGroups) {
    turns.sort((a, b) => (a.timestamp || '').localeCompare(b.timestamp || ''));
    const first = turns[0];

    // If this group is a pure carry-over (the previous batch already emitted a
    // record with uuid=promptId), suffix `:carry` so the DB holds two distinct
    // records that both get accumulated rather than one being deduplicated away.
    const baseUuid = (promptId === carryOverPromptId && !carryOverSeenInBatch)
      ? `${promptId}:carry`
      : promptId;

    // Label each turn with the file it writes (null if none)
    const turnFiles = turns.map(e => {
      const content = e.message?.content;
      if (!Array.isArray(content)) return null;
      for (const block of content) {
        if (block.type === 'tool_use' && WRITE_TOOLS.has(block.name)) {
          const fp = block.input?.file_path || block.input?.path;
          if (fp && typeof fp === 'string') return fp;
        }
      }
      return null;
    });

    const hasAnyFile = turnFiles.some(f => f !== null);

    if (!hasAnyFile) {
      // Pure conversation or Bash-only — one record, empty file_paths
      records.push(makeRecord(first, baseUuid, sumUsage(turns), []));
      continue;
    }

    // Attribute orphan turns (no file) to the NEXT anchor turn.
    // Trailing orphans after the last anchor → attributed to the last file seen.
    const fileGroups    = new Map();
    let lastFile        = turnFiles.find(f => f !== null);
    const firstAnchorIdx = turnFiles.findIndex(f => f !== null);

    for (let i = 0; i < turns.length; i++) {
      const file = turnFiles[i] !== null ? turnFiles[i]
        : (i < firstAnchorIdx ? turnFiles[firstAnchorIdx] : lastFile);

      if (!fileGroups.has(file)) fileGroups.set(file, []);
      fileGroups.get(file).push(turns[i]);
      if (turnFiles[i] !== null) lastFile = turnFiles[i];
    }

    // Emit one record per file; disambiguate uuid by appending file index
    let idx = 0;
    for (const [fp, fileTurns] of fileGroups) {
      const uuid = idx === 0 ? baseUuid : `${baseUuid}:${idx}`;
      records.push(makeRecord(first, uuid, sumUsage(fileTurns), [fp]));
      idx++;
    }
  }

  return records;
}

module.exports = { parseNewLines };
