// aicp-tracker/src/sender.js
'use strict';

const config = require('./config');
const wal    = require('./wal');

/**
 * Groups an array of raw usage records into the nested API payload:
 *   organization → useremail → date → [records]
 */
function buildPayload(cfg, records) {
  const org   = cfg.vcsUrl.replace(/^https?:\/\/[^/]+\//, '').replace(/\/+$/, ''); // extract slug, strip trailing slash
  const email = cfg.email;
  const payload = {};

  for (const rec of records) {
    const date = rec.timestamp ? rec.timestamp.slice(0, 10) : new Date().toISOString().slice(0, 10);
    payload[org]          = payload[org]          || {};
    payload[org][email]   = payload[org][email]   || {};
    payload[org][email][date] = payload[org][email][date] || [];
    payload[org][email][date].push({
      input_tokens:                     rec.input_tokens,
      cache_creation_input_tokens:      rec.cache_creation_input_tokens,
      cache_read_input_tokens:          rec.cache_read_input_tokens,
      output_tokens:                    rec.output_tokens,
      ephemeral_user_input_tokens:      rec.ephemeral_user_input_tokens,
      ephemeral_assistant_input_tokens: rec.ephemeral_assistant_input_tokens,
      ephemeral_tool_input_tokens:      rec.ephemeral_tool_input_tokens,
      web_search_requests:              rec.web_search_requests,
      web_fetch_requests:               rec.web_fetch_requests,
      service_tier:                     rec.service_tier,
      speed:                            rec.speed,
      model:                            rec.model,
      file_paths:                       rec.file_paths || [],
      enterprise_usd_per_token:         rec.enterprise_usd_per_token || null,
      sessionId:                        rec.sessionId,
      uuid:                             rec.uuid,
      parentUuid:                       rec.parentUuid,
      timestamp:                        rec.timestamp || null,
      gitBranch:                        rec.gitBranch,
      task_from_git_branch:             rec.task_from_git_branch || [],
      trace_uuids:                      rec.trace_uuids || '',
    });
  }

  return payload;
}

async function sendPending() {
  const pending = wal.getPending();
  if (!pending.length) return;

  let cfg;
  try { cfg = config.require(); } catch (e) { console.warn('[sender]', e.message); return; }

  const payload = buildPayload(cfg, pending);
  const uuids   = pending.map(r => r.uuid).filter(Boolean);

  try {
    const res = await fetch(`${cfg.apiUrl}/api/v1/ingest`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': cfg.apiKey,
      },
      body: JSON.stringify({ plan: cfg.plan, data: payload }),
      signal: AbortSignal.timeout(30_000),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`HTTP ${res.status}: ${text.slice(0, 200)}`);
    }

    wal.markSent(uuids);
    console.log(`[sender] Sent ${pending.length} records (${uuids.length} with UUIDs)`);
  } catch (e) {
    wal.markFailed(uuids);
    console.warn('[sender] Send failed (will retry next cycle):', e.message);
  }
}

module.exports = { sendPending };
