// aicp-tracker/src/wal.js
'use strict';

// Write-Ahead Log — records are written atomically to disk before any send attempt.
// Status: 'pending' | 'sent' | 'failed'
// On each cycle: retry all pending + failed, mark sent, prune old sent records.

const fs   = require('fs');
const path = require('path');
const { CONFIG_DIR } = require('./config');

const WAL_FILE  = path.join(CONFIG_DIR, 'wal.json');
const SENT_TTL  = 7 * 24 * 60 * 60 * 1000; // keep sent records for 7 days

function load() {
  try { return JSON.parse(fs.readFileSync(WAL_FILE, 'utf8')); } catch { return { records: [] }; }
}

function _save(wal) {
  const tmp = WAL_FILE + '.tmp';
  fs.writeFileSync(tmp, JSON.stringify(wal, null, 2), 'utf8');
  fs.renameSync(tmp, WAL_FILE); // atomic on POSIX; best-effort on Windows
}

function append(records) {
  const wal = load();
  const now = Date.now();
  for (const rec of records) {
    wal.records.push({ ...rec, _walStatus: 'pending', _walAddedAt: now, _walAttempts: 0 });
  }
  _save(wal);
}

function getPending() {
  const wal = load();
  return wal.records.filter(r => r._walStatus === 'pending' || r._walStatus === 'failed');
}

function markSent(uuids) {
  const set = new Set(uuids);
  const wal = load();
  const now = Date.now();
  for (const r of wal.records) {
    if (set.has(r.uuid)) { r._walStatus = 'sent'; r._walSentAt = now; }
  }
  // Prune records sent more than 7 days ago
  wal.records = wal.records.filter(r => !(r._walStatus === 'sent' && now - r._walSentAt > SENT_TTL));
  _save(wal);
}

function markFailed(uuids) {
  const set = new Set(uuids);
  const wal = load();
  for (const r of wal.records) {
    if (set.has(r.uuid)) { r._walStatus = 'failed'; r._walAttempts = (r._walAttempts || 0) + 1; }
  }
  _save(wal);
}

function pendingCount() {
  return getPending().length;
}

async function flush() {
  const count = pendingCount();
  if (count === 0) { console.log('WAL: nothing pending.'); return; }
  console.log(`WAL: ${count} pending records. Run \`aicp-tracker start\` to flush, or wait for next cycle.`);
}

module.exports = { append, getPending, markSent, markFailed, pendingCount, flush };
